gateway/src/apicast/configuration.lua

```diff
@@ -140,21 +142,47 @@ function _M.services_limit()
 end
 
 function _M.filter_services(services, subset)
-  subset = subset and util.to_hash(subset) or _M.services_limit()
-  if not subset or not next(subset) then return services end
+  local selected_services = {}
+  local service_regexp_filter  = env.value("APICAST_SERVICES_FILTER_BY_URL")
+  if service_regexp_filter == "" then
+      service_regexp_filter = default_regexp
+  end
+
+  if service_regexp_filter then
+    -- Checking that the regexp sent is correct, if not an empty service list
+    -- will be returned.
+    local _, err = match("", service_regexp_filter, 'oj')
+    if err then
+      -- @todo this return and empty list, Apicast will continue running maybe
+      -- process need to be stopped here.
+      ngx.log(ngx.ERR, "APICAST_SERVICES_FILTER_BY_URL cannot compile and all services are filtering out")
+      return selected_services
+    end
+  end
 
-  local s = {}
+  subset = subset and util.to_hash(subset) or _M.services_limit()
+  if (not subset or not next(subset)) and not service_regexp_filter then return services end
+  subset = subset or {}
 
   for i = 1, #services do
     local service = services[i]
+    local success = false
```

Comment:
I think that after extracting `match_host` this can be simplified to something like `local success/choose = service:match_host(service_regexp_filter) or subset[service.id]`. What do you think?